File: examples/elm-core/src/basics.erl
-lang([erl2, st]).
-module(basics).
-compile(export_all).

%% MATHEMATICS

-spec add(integer(), integer()) -> integer().
add(X1, X2) -> X1 + X2.

-spec add(integer()) -> fun((integer()) -> integer()).
add(X1) -> fun(X2) -> add(X1, X2) end.

-spec sub(integer(), integer()) -> integer().
sub(X1, X2) -> X1 - X2.

-spec sub(integer()) -> fun((integer()) -> integer()).
sub(X1) -> fun(X2) -> sub(X1, X2) end.

-spec mul(integer(), integer()) -> integer().
mul(X1, X2) -> X1 * X2.

-spec mul(integer()) -> fun((integer()) -> integer()).
mul(X1) -> fun(X2) -> mul(X1, X2) end.

-spec idiv(integer(), integer()) -> integer().
idiv(X1, X2) -> X1 div X2.

-spec idiv(integer()) -> fun((integer()) -> integer()).
idiv(X1) -> fun(X2) -> idiv(X1, X2) end.

%% EQUALITY

-spec eq(A, A) -> boolean().
eq(X1, X2) -> X1 == X2.

-spec eq(A) -> fun((A) -> boolean()).
eq(X1) -> fun(X2) -> eq(X1, X2) end.

-spec neq(A, A) -> boolean().
neq(X1, X2) -> X1 =/= X2.

-spec neq(A) -> fun((A) -> boolean()).
neq(X1) -> fun(X2) -> neq(X1, X2) end.

-spec lt(A, A) -> boolean().
lt(X1, X2) -> X1 < X2.

-spec lt(A) -> fun((A) -> boolean()).
lt(X1) -> fun(X2) -> lt(X1, X2) end.

-spec gt(A, A) -> boolean().
gt(X1, X2) -> X1 > X2.

-spec gt(A) -> fun((A) -> boolean()).
gt(X1) -> fun(X2) -> gt(X1, X2) end.

-spec le(A, A) -> boolean().
le(X1, X2) -> X1 =< X2.

-spec le(A) -> fun((A) -> boolean()).
le(X1) -> fun(X2) -> le(X1, X2) end.

-spec ge(A, A) -> boolean().
ge(X1, X2) -> X1 >= X2.

-spec ge(A) -> fun((A) -> boolean()).
ge(X1) -> fun(X2) -> ge(X1, X2) end.

-spec min(A, A) -> A.
min(X1, X2) ->
    case lt(X1, X2) of
        true -> X1;
        false -> X2
    end.

-spec min(A) -> fun((A) -> A).
min(X1) -> fun(X2) -> basics:min(X1, X2) end.

-spec max(A, A) -> A.
max(X1, X2) ->
    case gt(X1, X2) of
        true -> X1;
        false -> X2
    end.

-spec max(A) -> fun((A) -> A).
max(X1) -> fun(X2) -> basics:max(X1, X2) end.

-enum order() :: lt{} | eq {} | gt{}.

-spec compare(A, A) -> order().
compare(X1, X2) ->
    case lt(X1, X2) of
        true -> lt{};
        false ->
            case eq(X1, X2) of
                true -> eq{};
                false -> gt{}
            end
    end.

-spec compare(A) -> fun((A) -> order()).
compare(X1) -> fun(X2) -> compare(X1, X2) end.

%% BOOLEANS

-spec 'not'(boolean()) -> boolean().
'not'(B) -> not B.

-spec 'and'(boolean(), boolean()) -> boolean().
'and'(B1, B2) -> B1 and B2.

-spec 'and'(boolean()) -> fun((boolean()) -> boolean()).
'and'(B1) -> fun(B2) -> 'and'(B1, B2) end.

-spec 'or'(boolean(), boolean()) -> boolean().
'or'(B1, B2) -> B1 or B2.

-spec 'or'(boolean()) -> fun((boolean()) -> boolean()).
'or'(B1) -> fun(B2) -> 'or'(B1, B2) end.

-spec 'xor'(boolean(), boolean()) -> boolean().
'xor'(B1, B2) -> B1 xor B2.

-spec 'xor'(boolean()) -> fun((boolean()) -> boolean()).
'xor'(B1) -> fun(B2) -> 'xor'(B1, B2) end.

%% FANCIER MATH

-spec mod_by(integer(), integer()) -> integer().
mod_by(X1, X2) -> X2 div X1.

-spec mod_by(integer()) -> fun((integer()) -> integer()).
mod_by(X1) -> fun(X2) -> mod_by(X1, X2) end.

-spec remainder_by(integer(), integer()) -> integer().
remainder_by(X1, X2) -> X2 rem X1.

-spec remainder_by(integer()) -> fun((integer()) -> integer()).
remainder_by(X1) -> fun(X2) -> remainder_by(X1, X2) end.

-spec negate(integer()) -> integer().
negate(X) -> -X.

-spec abs(integer()) -> integer().
abs(X) ->
    case lt(X, 0) of
        true -> -X;
        false -> X
    end.

-spec clamp(integer(), integer(), integer()) -> integer().
clamp(Low, High, Num) ->
    case lt(Num, Low) of
        true -> Low;
        false -> case gt(Num, High) of true -> High; false -> Num end
    end.

%% FUNCTION HELPERS

-spec composeL(fun((B) -> C), fun((A) -> B)) -> fun((A) -> C).
composeL(G, F) ->
    fun(X) -> G(F(X)) end.

-spec composeL(fun((B) -> C)) -> fun((fun((A) -> B)) -> fun((A) -> C)).
composeL(G) -> fun(F) -> composeL(G, F) end.

-spec composeR(fun((A) -> B), fun((B) -> C)) -> fun((A) -> C).
composeR(F, G) ->
    fun(X) -> G(F(X)) end.

composeR(F) ->
    fun(G) -> composeR(F, G) end.

-spec apR(A, fun((A) -> B)) -> B.
apR(X, F) ->
    F(X).

-spec apR(A) -> fun((fun((A) -> B)) -> B).
apR(X) -> fun(F) -> apR(X, F) end.

-spec apL(fun((A) -> B), A) -> B.
apL(F, X) ->
    F(X).

-spec apL(fun((A) -> B)) -> fun((A) -> B).
apL(F) -> fun(X) -> apL(F, X) end.

-spec identity(A) -> A.
identity(X) -> X.

-spec always(A, _) -> A.
always(A, _) -> A.

-spec always(A) -> fun((_) -> A).
always(A) -> fun(X) -> always(A, X) end.

-enum never() :: just_one_more{never()}.

-spec never(never()) -> _.
never(just_one_more{Nvr}) -> never(Nvr).

File: examples/elm-core/ir-spec/basics.ml
type order'0 = order

and order = Order'Lt | Order'Eq | Order'Gt

and never'0 = never

and never = Never'Just_one_more of never'0

let rec add'2 : int * int -> int = function v_X1, v_X2 -> v_X1 + v_X2

let rec add'1 : int -> int -> int = function
  | v_X1 -> ( function v_X2 -> add'2 (v_X1, v_X2) )

let rec sub'2 : int * int -> int = function v_X1, v_X2 -> v_X1 - v_X2

let rec sub'1 : int -> int -> int = function
  | v_X1 -> ( function v_X2 -> sub'2 (v_X1, v_X2) )

let rec mul'2 : int * int -> int = function v_X1, v_X2 -> v_X1 * v_X2

let rec mul'1 : int -> int -> int = function
  | v_X1 -> ( function v_X2 -> mul'2 (v_X1, v_X2) )

let rec idiv'2 : int * int -> int = function v_X1, v_X2 -> v_X1 / v_X2

let rec idiv'1 : int -> int -> int = function
  | v_X1 -> ( function v_X2 -> idiv'2 (v_X1, v_X2) )

let rec eq'2 : 'tA * 'tA -> bool = function v_X1, v_X2 -> v_X1 = v_X2

let rec eq'1 : 'tA -> 'tA -> bool = function
  | v_X1 -> ( function v_X2 -> eq'2 (v_X1, v_X2) )

let rec neq'2 : 'tA * 'tA -> bool = function v_X1, v_X2 -> v_X1 <> v_X2

let rec neq'1 : 'tA -> 'tA -> bool = function
  | v_X1 -> ( function v_X2 -> neq'2 (v_X1, v_X2) )

let rec lt'2 : 'tA * 'tA -> bool = function v_X1, v_X2 -> v_X1 < v_X2

let rec lt'1 : 'tA -> 'tA -> bool = function
  | v_X1 -> ( function v_X2 -> lt'2 (v_X1, v_X2) )

let rec gt'2 : 'tA * 'tA -> bool = function v_X1, v_X2 -> v_X1 > v_X2

let rec gt'1 : 'tA -> 'tA -> bool = function
  | v_X1 -> ( function v_X2 -> gt'2 (v_X1, v_X2) )

let rec le'2 : 'tA * 'tA -> bool = function v_X1, v_X2 -> v_X1 <= v_X2

let rec le'1 : 'tA -> 'tA -> bool = function
  | v_X1 -> ( function v_X2 -> le'2 (v_X1, v_X2) )

let rec ge'2 : 'tA * 'tA -> bool = function v_X1, v_X2 -> v_X1 >= v_X2

let rec ge'1 : 'tA -> 'tA -> bool = function
  | v_X1 -> ( function v_X2 -> ge'2 (v_X1, v_X2) )

let rec min'2 : 'tA * 'tA -> 'tA = function
  | v_X1, v_X2 -> (
      match lt'2 (v_X1, v_X2) with true -> v_X1 | false -> v_X2 )

let rec min'1 : 'tA -> 'tA -> 'tA = function
  | v_X1 -> ( function v_X2 -> min'2 (v_X1, v_X2) )

let rec max'2 : 'tA * 'tA -> 'tA = function
  | v_X1, v_X2 -> (
      match gt'2 (v_X1, v_X2) with true -> v_X1 | false -> v_X2 )

let rec max'1 : 'tA -> 'tA -> 'tA = function
  | v_X1 -> ( function v_X2 -> max'2 (v_X1, v_X2) )

let rec compare'2 : 'tA * 'tA -> order'0 = function
  | v_X1, v_X2 -> (
      match lt'2 (v_X1, v_X2) with
      | true -> Order'Lt
      | false -> (
          match eq'2 (v_X1, v_X2) with true -> Order'Eq | false -> Order'Gt ) )

let rec compare'1 : 'tA -> 'tA -> order'0 = function
  | v_X1 -> ( function v_X2 -> compare'2 (v_X1, v_X2) )

let rec not'1 : bool -> bool = function v_B -> not v_B

let rec and'2 : bool * bool -> bool = function v_B1, v_B2 -> v_B1 && v_B2

let rec and'1 : bool -> bool -> bool = function
  | v_B1 -> ( function v_B2 -> and'2 (v_B1, v_B2) )

let rec or'2 : bool * bool -> bool = function v_B1, v_B2 -> v_B1 || v_B2

let rec or'1 : bool -> bool -> bool = function
  | v_B1 -> ( function v_B2 -> or'2 (v_B1, v_B2) )

let rec xor'2 : bool * bool -> bool = function
  | v_B1, v_B2 -> (not v_B1) <> not v_B2

let rec xor'1 : bool -> bool -> bool = function
  | v_B1 -> ( function v_B2 -> xor'2 (v_B1, v_B2) )

let rec mod_by'2 : int * int -> int = function v_X1, v_X2 -> v_X2 / v_X1

let rec mod_by'1 : int -> int -> int = function
  | v_X1 -> ( function v_X2 -> mod_by'2 (v_X1, v_X2) )

let rec remainder_by'2 : int * int -> int = function
  | v_X1, v_X2 -> v_X2 mod v_X1

let rec remainder_by'1 : int -> int -> int = function
  | v_X1 -> ( function v_X2 -> remainder_by'2 (v_X1, v_X2) )

let rec negate'1 : int -> int = function v_X -> -v_X

let rec abs'1 : int -> int = function
  | v_X -> ( match lt'2 (v_X, 0) with true -> -v_X | false -> v_X )

let rec clamp'3 : int * int * int -> int = function
  | v_Low, v_High, v_Num -> (
      match lt'2 (v_Num, v_Low) with
      | true -> v_Low
      | false -> (
          match gt'2 (v_Num, v_High) with true -> v_High | false -> v_Num ) )

let rec composeL'2 : ('tB -> 'tC) * ('tA -> 'tB) -> 'tA -> 'tC = function
  | v_G, v_F -> ( function v_X -> v_G (v_F v_X) )

let rec composeL'1 : ('tB -> 'tC) -> ('tA -> 'tB) -> 'tA -> 'tC = function
  | v_G -> ( function v_F -> composeL'2 (v_G, v_F) )

let rec composeR'2 : ('tA -> 'tB) * ('tB -> 'tC) -> 'tA -> 'tC = function
  | v_F, v_G -> ( function v_X -> v_G (v_F v_X) )

let rec composeR'1 : ('tA -> 'tB) -> ('tB -> 'tC) -> 'tA -> 'tC = function
  | v_F -> ( function v_G -> composeR'2 (v_F, v_G) )

let rec apR'2 : 'tA * ('tA -> 'tB) -> 'tB = function v_X, v_F -> v_F v_X

let rec apR'1 : 'tA -> ('tA -> 'tB) -> 'tB = function
  | v_X -> ( function v_F -> apR'2 (v_X, v_F) )

let rec apL'2 : ('tA -> 'tB) * 'tA -> 'tB = function v_F, v_X -> v_F v_X

let rec apL'1 : ('tA -> 'tB) -> 'tA -> 'tB = function
  | v_F -> ( function v_X -> apL'2 (v_F, v_X) )

let rec identity'1 : 'tA -> 'tA = function v_X -> v_X

let rec always'2 : 'tA * _ -> 'tA = function v_A, _ -> v_A

let rec always'1 : 'tA -> _ -> 'tA = function
  | v_A -> ( function v_X -> always'2 (v_A, v_X) )

let rec never'1 : never'0 -> _ = function
  | Never'Just_one_more v_Nvr -> never'1 v_Nvr
